Add Spaces data model and operations. (#3)
did-coop/wallet-attached-storage-spec: spec.md

### Space Data Model

`Space` properties:

* `id` - deterministically set by server if not provided.

> * `type` - set by server to the string `Space`

Reviewer: should we allow `['{uri}', 'Space']`? (imo yes)

(I'd also be willing to constrain and say these arrays must be sorted)

Author: ohhh right, arrays. Yeah one sec...

Author: Updated

### Space Data Model

`Space` properties:

* `id` - deterministically set by server if not provided.
* `type` - set by server to the string `Space`
* `name` (optional)
* `controller` - determined by server based on who creates the space originally

#### Space JSON Representation

#### Space CBOR Representation

### Create Space operation

To create a Space:

> * Perform an authenticated Create Space operation. The controller/invoker DID

Reviewer: (not urgent)

can invoker be different than controller, but with a verifiable zcap delegation chain authorizing POST /spaces/ ?
(yes imho)

Author: Agreed, yes. We'll add that clarification when we explain delegations.

Author: I'll add a placeholder for that tho

  of the operation is set as the Space's `controller`

> * If an `id` is provided in the Create Space request, it must be a `urn:uuid`.

Reviewer: ```suggestion
* If an `id` is provided in the Create Space request, it must start with `urn:uuid`.
```

Content-type: application/json
Authorization: ...

{
  "name": "Example space #1"
}
```

Example success response:

```http
HTTP/1.1 201 Created
Content-type: application/json
Location: https://example.com/spaces/81246131-69a4-45ab-9bff-9c946b59cf2e

{
  "id": "81246131-69a4-45ab-9bff-9c946b59cf2e",
  "type": ["Space"],
  "name": "Example space #1",
  "controller": "did:key:z6MkpBMbMaRSv5nsgifRAwEKvHHoiKDMhiAHShTFNmkJNdVW"
}
```

Example error response (missing authorization, unable to determine controller):

Example error response (invalid authorization):

Example error response (missing or insufficient onboarding material provided):

Example error response (invalid `id` provided):

Example error response (a space with the specified `id` already exists):

#### (Javascript API) Space .create()

### Read Space operation

* Requires appropriate authorization (root zcap invoked by the space's controller,
  or a zcap granting permission to read a particular space)
* Returns the details for the specified space `id`

#### (HTTP API) GET `/spaces/{space_id}`

Example request:

```http
GET /spaces/81246131-69a4-45ab-9bff-9c946b59cf2e HTTP/1.1
Host: example.com
Accept: application/json

> Authorization: ...

Reviewer: ```suggestion
Authorization: Signature keyId="did:key:z6MkpBMbMaRSv5nsgifRAwEKvHHoiKDMhiAHShTFNmkJNdVW#z6MkpBMbMaRSv5nsgifRAwEKvHHoiKDMhiAHShTFNmkJNdVW" ...
```

Up to you whether you want this to ugly up the example, but imho it can help illustrate the auth

Author: oh absolutely, I figured we'll come back and add the actual headers

```

Example success response:

```http
HTTP/1.1 200 OK
Content-type: application/json

{
  "id": "81246131-69a4-45ab-9bff-9c946b59cf2e",
  "type": ["Space"],
  "name": "Example space #1",
  "controller": "did:key:z6MkpBMbMaRSv5nsgifRAwEKvHHoiKDMhiAHShTFNmkJNdVW"
}
```

Example error response (missing or insufficient authorization):

Example error response (space id not found):

> #### (HTTP API) HEAD `/spaces/{space_id}`

Reviewer: (where we also consider `OPTIONS`/CORS)

> #### (HTTP API) HEAD `/spaces/{space_id}`

Author: oh I was gonna add OPTIONS in addition to HEAD. since the semantics are different. Also, no CORS (CORS is the devil, just ask Tim)

> #### (HTTP API) HEAD `/spaces/{space_id}`

Author: But seriously, Tim (correctly) argues is that CORS is a band-aid for when you dont' have proper authorization. and we do

> #### (HTTP API) HEAD `/spaces/{space_id}`

Reviewer: @dmitrizagidulin ok but for practical reasons if we support a little CORS, it will make a data pub server able to receive cross-domain `fetch` which seems very useful

> #### (HTTP API) HEAD `/spaces/{space_id}`

Author: @gobengo oh, agreed, yeah.

```http
HTTP/1.1 200 OK
Content-type: application/json

...
```

Example success response (requester does NOT have access to any spaces):

```http
HTTP/1.1 200 OK
Content-type: application/json

...
```

Example error response (missing authorization):

### Update Space operation

* Requires appropriate authorization (root zcap invoked by the space's controller,
  or a zcap granting permission to write to a particular space)
* Allows to update the following fields:
  - `name`
  - `controller`

#### (HTTP API) PUT `/spaces/{space_id}`

Note that this is a _full_ update (partial updates via http `PATCH` verb might
be supported later). However, some fields may not be updated (like `id`) and so
may be omitted from the request payload.

Note that this operation is idempotent.

Example request (updating the name of a space):

```http
PUT /spaces/81246131-69a4-45ab-9bff-9c946b59cf2e HTTP/1.1
Host: example.com
Content-type: application/json
Accept: application/json
Authorization: ...

{
  "id": "81246131-69a4-45ab-9bff-9c946b59cf2e",
  "name": "Newly renamed space #1",
  "controller": "did:key:z6MkpBMbMaRSv5nsgifRAwEKvHHoiKDMhiAHShTFNmkJNdVW"
}
```

> Example success response (returns the updated details of the space):

Reviewer: Can this be a 204? I dont want to require the server to pay the bandwidth here. The client can opt into it by sending a GET

Author: Sure. It's a minor pedantic nicety (to return the new state after PUT), but you make a good point about saving bandwidth

  "type": ["Space"],
  "name": "Newly renamed space #1",
  "controller": "did:key:z6MkpBMbMaRSv5nsgifRAwEKvHHoiKDMhiAHShTFNmkJNdVW"
}
```

Example error response (missing or invalid authorization):

Example error response (invalid `id` provided, the space does not exist):

Example error response (client is attempting to change an immutable field like
the space `id`):

### Delete Space operation

* Requires appropriate authorization (root zcap invoked by the space's controller,
  or a zcap granting permission to write to a particular space)
* Deletes the space and all of the data (collections and resources) contained
  in it
* Server responds with a Tombstone object (which will be later used for replication)
* This operation is idempotent

#### (HTTP API) DELETE `/spaces/{space_id}`

Example request (no request body):

```http
DELETE /spaces/81246131-69a4-45ab-9bff-9c946b59cf2e HTTP/1.1
Host: example.com
Accept: application/json
Authorization: ...
```

Example success response:

```http
HTTP/1.1 200 OK
Content-type: application/json

{
  ... TODO: specify Tombstone format
}
```

TODO: Decide whether subsequent GET requests to the same space should result in

> a 410 Gone, or the usual 404.

Reviewer: imho 404 not 410.

Author: ok but then how else am I gonna prove my http hipster status (if I don't suggest 410)? 
oooh i know, I'll mint a self-issued VC giving me hipster status...

Author: (I joke, but there's arguments for 410, but we can discuss separately)
